### ISSUE TITLE
warn when default provider overrides custom provider.

### ISSUE THREAD

**Issue by robert-king** [angular/angular#51090]
### Description

I spent some time debugging why my custom ErrorHandler wasn't being provided, it turns out it was being overridden by a default provider further down in the provider list. It would be nice to have a warning in this situation.

### Proposed solution

n/a

### Alternatives considered

n/a

**Comment by JeanMeche** [comment]
You mean you had multiple providers for the same token in the providers list (without `multi:true` of course) ?

**Comment by robert-king** [comment]
Ah I didn't use multi:true, can you do that with error handler? Here are my providers fyi,
it works only if i put provide: ErrorHandler, near the bottom of the list.

providers: [
    {
      provide: HTTP_INTERCEPTORS,
      useClass: AuthHttpInterceptor,
      multi: true,
    },
    importProvidersFrom(
      AuthModule.forRoot({
        authorizationParams: {
          audience: environment.auth0Audience,
          // redirect_uri:
        },
        domain: environment.auth0Domain,
        clientId: environment.auth0ClientId,
        cacheLocation: 'localstorage',
        httpInterceptor: {
          allowedList: environment.auth0ApiAllowedList,
        },
      }),
    ),
    provideHttpClient(withInterceptorsFromDi()),
    provideRouter(appRoutes, withEnabledBlockingInitialNavigation()),
    importProvidersFrom(BrowserAnimationsModule),
    importProvidersFrom(MatDialogModule),
    importProvidersFrom(MatSnackBarModule),
    importProvidersFrom(MatNativeDateModule),
    {
      provide: CLIENT_TOKEN,
      useFactory: clientFactory,
      deps: [HttpClient],
    },
    {
      provide: MAT_FORM_FIELD_DEFAULT_OPTIONS,
      useValue: { appearance: 'fill', subscriptSizing: 'dynamic' },
    },
    {
      provide: ErrorHandler,
      useValue: Sentry.createErrorHandler({
        showDialog: false,
      }),
    },
    {
      provide: Sentry.TraceService,
      deps: [Router],
    },
    {
      provide: APP_INITIALIZER,
      useFactory: () => () => {},
      deps: [Sentry.TraceService],
      multi: true,
    },
    {
      provide: Sentry,
      useValue: Sentry,
    },

**Comment by alxhub** [comment]
So the provider using `Sentry.createErrorHandler` is the one that took effect? Which provider did it override?

**Comment by robert-king** [comment]
@alxhub  here's a repro:

https://stackblitz.com/edit/stackblitz-starters-swyiqg?file=src%2Fmain.ts

i.e.

```
bootstrapApplication(App, {
  providers: [
    {
      provide: ErrorHandler,
      useClass: MyErrorHandler,
    },
    importProvidersFrom(BrowserAnimationsModule),
  ],
});
```

I guess BrowserAnimationsModule overwrites it. Was a bit confusing for me.

**Comment by JeanMeche** [comment]
Yes, `BrowserAnimationsModule` exports the `BrowserModule` which sets up the default ErrorHandler provider.

**Comment by robert-king** [comment]
Yeah that's apparent from my simple example, but it's not obvious or beginner friendly, hence why I was suggesting a warning. Another option could be to update the docs.

https://angular.io/api/core/ErrorHandler
https://angular.io/api/platform-browser/BrowserModule

**Comment by alxhub** [comment]
Oooh, okay. That's super interesting.

I don't think we can have an error like you suggest, because overriding a provider is a very normal operation. However, the real issue here seems to be that `BrowserAnimationsModule` does more than it advertises and tries to configure other providers that aren't related to animations.

**Comment by robert-king** [comment]
yeah i wasn't expecting BrowserAnimationsModule to do that.

A warning, not an error. And the warning could be option.

with routing, there is a "withDebugging()" option.
perhaps there could be something similar e.g.

{
      provide: ErrorHandler,
      useClass: MyErrorHandler,
      withDebugging: True
    },
